File: pyplayground/client/RobotEPuck.py
"""
Wrapper para el robot del tipo EPuck

    from pyplayground.client import RobotFactory

    rob = RobotFactory.connectRobot( 'Epuck-01', host, port )
    rob.setSpeed( -1000, 1000 )
    sensores = rob.getSensors()
    rob.setLedRing( True )
    image = rob.getCameraImage()
"""
import socket

import pyplayground.utils.BasicSockJson as BasicSockJson
import pyplayground.client.RobotBase as RobotBase

class RobotEPuck( RobotBase.RobotBase ):
    def __init__( self, name:str, host:str, port:int, sock:socket.socket ):
        super().__init__( name, host, port, sock )

    def __str__( self ):
        return f'RobotEnki >> name:{self.name} - host={self.host} - port={self.port}'

    def setLedRing( self, on_off:bool ) -> None:
        """
        apaga o enciende el anillo que rodea al robot

        Parameters
          on_off: True enciende el anillo, False lo apaga

        """
        led_on = 1 if on_off else 0
        pkg = { 'cmd':'setLedRing', 'estado': on_off }
        BasicSockJson.send( self.sock, pkg )
        resp = BasicSockJson.read( self.sock, self.buff )['answer']
        return None

    def getCameraImage( self ) -> list :
        """
        Obtiene la imagen de la camara lineal del robot

        Return
            una lista con los puntos de la imagen (60 de ancho x 1 de alto)
        """
        pkg = { 'cmd':'getCameraImage' }
        BasicSockJson.send( self.sock, pkg )
        resp = BasicSockJson.read( self.sock, self.buff )['answer']['image']
        resp = bytearray( resp, 'iso-8859-1' )
        l = len( resp )
        resp = [ tuple( resp[i:i+4] ) for i in range( 0, l, 4 ) ]
        return resp

File: java/src/main/java/rcr/jplayground/RobotEPuck.java
package rcr.jplayground;

import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import org.json.*;

public class RobotEPuck extends RobotBase {

    public RobotEPuck( String name, String host, int port ) throws Exception {
        super( name, host, port, "epuck" );
    }

    public void setLedRing( boolean on_off ) throws Exception {

        JSONObject json = new JSONObject();
        json.put( "cmd", "setLedRing" );
        json.put( "estado", on_off ? 1 : 0 );
        out.println( json );

        String resp = in.readLine();
    }

    public void getSensors() throws Exception {
        Map<String, Object> map = retrieveSensors();
    }

    public List<int[]> getCameraImage() throws Exception {
        JSONObject json = new JSONObject();
        json.put( "cmd", "getCameraImage" );
        out.println( json );

        int[] s = new int[4];
        for( int i = 0; i<4; i++ )
            s[i] = in.read();

        int len = (s[0]<<24) + (s[1]<<16) + (s[2]<<8) + s[3];

        int[] data = new int[len];
        for( int i = 0; i<len; i++ )
            data[i] = in.read();

        List<int[]> imagen = new ArrayList<>();
        for( int i=0; i<len; i+= 4 ) {
            int[] color = { data[i+0], data[i+1], data[i+2], data[i+3] };
            imagen.add( color );
        }

        return imagen;
    }

}
